BOLT 12: re-add recurrence support.

rustyrussell opened the pull request:
This is taken from the old draft, and updated with modern field numbers and names.

I've made some minor changes:
1. The `proportional_amount` field comes before `seconds_before` which seems more logical to me.
2. The `offer_absolute_expiry` is clarified to only apply to the *initial* invoice request: once you're recurring the other fields take over.

The complexities mainly come from two sources:

1. A requirement to be precise about when recurrence happens.  This is because we want a push system, not a pull, since only the former allows for real authorization by the user, and there are many different reasonable ways to do recurrence.
2. This is the first wallet feature which isn't just "fire and forget", but requires them to keep state and do unprompted actions.

One proposal I would like to add is a courtesy onion message when a user cancels a periodic payment.  This is generally useful to tell the difference between a deliberate cancellation and a user-related failure.

rustyrussell committed:
BOLT 12: re-add recurrence support.

This is taken from the old draft, and updated with modern field numbers and names.

I've made some minor changes:
1. The `proportional_amount` field comes before `seconds_before` which seems more logical to me.
2. The `offer_absolute_expiry` is clarified to only apply to the *initial* invoice request: once you're recurring the other fields take over.

The complexities mainly come from two sources:

1. A requirement to be precise about when recurrence happens.  This is because we want a push system, not a pull, since only the former allows for real authorization by the user, and there are many different reasonable ways to do recurrence.
2. This is the first wallet feature which isn't just "fire and forget", but requires them to keep state and do unprompted actions.

One proposal I would like to add is a courtesy onion message when a user cancels a periodic payment.  This is generally useful to tell the difference between a deliberate cancellation and a user-related failure.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,88 @@ The human-readable prefix for offers is `lno`.
+   * which repeat from whenever you made the first payment
+
+Thus, each offer containing a recurring payment has:
+1. A `time_unit` defining 0 (seconds), 1 (days), 2 (months), 3 (years).
```
I understand the need for seconds, days and months but years seem redundant as 1 year is always 12 months.

rustyrussell replied:
True.  I could argue that 4 years is better UX than 48 months, but the same argument would apply for minutes and hours.  I'll drop it!

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,88 @@ The human-readable prefix for offers is `lno`.
+3. An optional `offer_recurrence_limit` of total payments to be paid.
+4. An optional `offer_recurrence_base`:
+   * `basetime`, defining when the first period starts in seconds since 1970-01-01 UTC.
+   * `start_any_period` if non-zero, meaning you don't have to start paying at the period indicated by `basetime`, but can use `offer_recurrence_start` to indicate what period you are starting at.
```
```suggestion
   * `start_any_period` if non-zero, meaning you don't have to start paying at the period indicated by `basetime`, but can use `invreq_recurrence_start` to indicate what period you are starting at.
```

thomash-acinq replied:
Do we ever want to force the payer to pay for all the previous periods? It seems to me that `start_any_period` should always be non-zero.

rustyrussell replied:
No, it would mean if you don't start at the beginning you can't join at all.

But perhaps we should rely on the expiration of the offer for this, instead?

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,88 @@ The human-readable prefix for offers is `lno`.
+  - calculate the offset in days within the month of period #0 start.
+  - calculate the offset in seconds within the day of period #0 start.
+  - add `period` multiplied by `N` months to get the month of the period start.
+  - add the offset days to get the day of the period start.
+    - if the day is not within the month, use the last day within the month.
+  - add the offset seconds to get the period start in seconds.
```
If the day is not within the month, we should use the last day of the month **and the last second of that day**.

We want a one month period starting on January 30th at 2AM to end before a one month period starting on January 31th at 1AM as it starts almost a full day later. However with your proposal, the former would end an hour after the latter.

rustyrussell replied:
If you're having monthly recurrences, you don't care about +/- a day,  BUT we have to define it somehow.  I think you'll find this easier to implement as written, so I think that's better.

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,88 @@ The human-readable prefix for offers is `lno`.
+  - add the offset seconds to get the period start in seconds.
+- otherwise, the time is invalid.
+
+Note that offset seconds can overflow only if the period start is in a leap second; we ignore this!
```
If we ignore leap seconds, then there is no need for days as a time unit as seconds can be used instead.

rustyrussell replied:
We don't ignore, we simply ignore the case where period start *is on* a leap second.  We should probably round it down, for consistency in that case?

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -282,6 +364,41 @@ A writer of an offer:
       - MUST set `offer_quantity_max` to 0.
   - otherwise:
     - MUST NOT set `offer_quantity_max`.
+  - MAY include `offer_recurrence` to indicate offer should trigger time-spaced invoices.
```
```suggestion
  - MAY include `offer_recurrence` to indicate offer should trigger time-spaced invoice requests.
```

thomash-acinq commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,88 @@ The human-readable prefix for offers is `lno`.
     1. type: 22 (`offer_issuer_id`)
     2. data:
         * [`point`:`id`]
+    1. type: 24 (`offer_recurrence`)
```
I think we should use odd TLVs so that a payer that doesn't support recurrence can pay for one period.

rustyrussell replied:
I wondered about that!  I think that may be worse UX, depending on the use case?  They will think it's a one-off payment ("hey that's cheap!") and then get upset when it's not.

rustyrussell replied:
Ok, now I'm using both.  That way the user can decide...

a-mpch commented on 12-offer-encoding.md:
```diff
@@ -348,6 +473,8 @@ painful to have to special-case the "only one left" offer generation.
 
 Offers can be used to simply send money without expecting anything in return (tips, kudos, donations, etc), which means the description field is optional (the `offer_issuer` field is very useful for this case!); if you are charging for something specific, the description is vital for the user to know what it was they paid for.
 
+We insist that recurring requests be in order (thus, if you pay an invoice for #34 of a recurring offer, it implicitly commits to the successful payment of #0 through #33).  The `invreq_recurrence_paywindow` constrains how far you can pay in advance precisely, and if it isn't in the offer the defaults provide some slack, without allowing commitments into the far future.
```
I think `invreq_recurrence_paywindow` is a typo to `offer_recurrence_paywindow`.

if not is missing in the TLVs and I don't get how it would work 👀 . `invreq_recurrence_paywindow` would be set by the payer but it would be controlled by the issuer

rustyrussell replied:
Good catch.

a-mpch replied:
@rustyrussell I think this one might slipped out from your changes

rustyrussell replied:
"This time for sure!"

rustyrussell commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,88 @@ The human-readable prefix for offers is `lno`.
+   * `start_any_period` if non-zero, meaning you don't have to start paying at the period indicated by `basetime`, but can use `offer_recurrence_start` to indicate what period you are starting at.
+5. An optional `offer_recurrence_paywindow`:
+   * `seconds_before`, defining how many seconds prior to the start of the period a payment will be accepted.
+   * `proportional_amount`, if set indicating that a payment made during the period itself will be charged proportionally to the remaining time in the period (e.g. 150 seconds into a 1500 second period gives a 10% discount).
```
Indeed.  Let's give it a separate TLV entry.

rustyrussell replied:
Actually, it belongs in "offer_recurrence_base".  If that is not present, the period starts when you request the invoice, so it's always "the entire period".

rustyrussell committed:
BOLT12 recurrence: feedback from ThomasH and  a-mpch

1. Remove start_any_period.  If you want them not to start in a random
   period (only appliciable if you set `offer_recurrence_base`) then
   use offer_absolute_expiry.
2. Remove years, use months.
3. Clarify leap seconds.
4. Put `proportional_amount` inside `offer_recurrence_base` where it
   belongs, not inside paywindow.
   
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: renumber recurrence fields.

Only offer_recurrence_base is compulsory, really.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell commented:
OK, updated.  I will need to rework the test vectors, however!

rustyrussell committed:
BOLT 12: Allow either optional or compulsory recurrence variants.

This allows the writer of the offer to decide how
pre-recurrence-supporting wallets should behave.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

vincenzopalazzo reviewed (commented):
LGTM overall, I want to take another look later in the month, but probably in this PR we can also remove the FIXME N 7 in the https://github.com/lightning/bolts/blob/master/12-offer-encoding.md#fixme-possible-future-extensions

GBKS commented:
> One proposal I would like to add is a courtesy onion message when a user cancels a periodic payment. This is generally useful to tell the difference between a deliberate cancellation and a user-related failure.

Subscriptions likely require more communication. From the initial sharing of the offer, to price changes, to notifying a user about a missed payment, providing a receipt, etc. Maybe best to leave this type of communication up to the participants to handle in their medium of choice?

sbddesign commented:
> Subscriptions likely require more communication. From the initial sharing of the offer, to price changes, to notifying a user about a missed payment, providing a receipt, etc. Maybe best to leave this type of communication up to the participants to handle in their medium of choice?

@GBKS I agree that from a UX perspective there is always going to be more communication required between the subscriber and the recipient. However, if there is a protocol-native way to _create_ a subscription, I think there also needs to be a protocol-native way to _cancel_ the subscription.

GBKS commented:
> However, if there is a protocol-native way to create a subscription, I think there also needs to be a protocol-native way to cancel the subscription.

Maybe. You're creating a subscription proposal, but it's not active until the payee accepts it and acts on it. It's more like a subscription request, like we have payment requests. So if you want a "subscription cancelled" message, you may also want a "subscription accepted" message. But TBH, I don't have a super strong opinion. Just doesn't seem 100% right to exclusively have the cancel message.

rustyrussell commented:
Note: @TheBlueMatt pointed out that the correct behavior is to send the courtesy "cancel" message in place of the next request.  Slightly more complex to implement (your wallet now needs to remember cancelled subscriptions), but it's a nice balance between letting the vendor know, and not letting them cut you off early.  Apparently Apple Pay does something similar, and I can see the logic.

From an implementation POV, this may simply look like a stub invoice_request?

rustyrussell commented:
> > One proposal I would like to add is a courtesy onion message when a user cancels a periodic payment. This is generally useful to tell the difference between a deliberate cancellation and a user-related failure.
> 
> Subscriptions likely require more communication. From the initial sharing of the offer, to price changes, to notifying a user about a missed payment, providing a receipt, etc. Maybe best to leave this type of communication up to the participants to handle in their medium of choice?

That, however, has proven to be a centralized disaster, and severely limits interoperation.  You cannot automate this unspecified stuff...

Eventually we will probably want offer replacement (where we reject an invoice_request and optionally refers to an updated offer), but one step at a time!

rustyrussell committed:
BOLT 12: really fix `invreq_recurrence_paywindow` typo.

Thanks https://github.com/a-mpch!

Also replace some errant tabs.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: Add recurrence cancel invreq.

This is a best-effort message, but helps differentiate a deliberate
cancellation by a user from wallet issues.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell commented:
Added the "cancel" flag.  You basically send an invreq like normal, but it says "cancel" instead of "send me the next invoice".

Also added the requirement to fetch the next invoice explicitly to the requirements for recurring payments!

rustyrussell committed:
BOLT 12: use subtypes for duplicated non-trivial fields.

Recurrence has the same fields in offer, invoice_request and invoice, and bip_353_name
has the same fields in invoice_request and invoice.

This is slightly clearer, and it makes code generation cleaner (since
they are clearly the same type, they are easier to use).

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT12: more cleanups (from implementing).

1. Fix indent for `invreq_recurrence_cancel`
2. Don't say you can omit `invreq_paths` since you would anyway.
3. Don't allow setting `invreq_recurrence_cancel` in the "not responding to an offer" case.
4. Put the "no future invoices" case under the invoice writer section where it bellongs.
5. Explicitly allow skipping invreq_quantity and invreq_amount on the reader side if we're cancelling (we said you could on the writer side).
6. Add missing check for offer_absolute_expiry on the invoice writer side.
7. Make it clear that you can't issue an invoice if it's already been paid.
8. Disallow `invreq_recurrence_cancel` if not a recurring invoice.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

Ramin078 reviewed (commented):
rustyrussell:guilt/offers-recurrence

Ramin078 commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+
+Thus, each offer containing a recurring payment has:
+1. A `time_unit` defining 0 (seconds), 1 (days), 2 (months).
+2. A `period`, defining how often (in `time_unit`) it has to be paid.
```
raminasadi009@gmail.com

Ramin078 reviewed (commented):
rustyrussell:guilt/offers-recurrence 69099

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+        * [`recurrence_paywindow`:`recurrence_paywindow`]
+    1. type: 29 (`offer_recurrence_limit`)
+    2. data:
+        * [`tu32`:`max_period`]
```
`max_period` indicates the maximum number of periods to trigger/pay the offer, but could the name be mistaken for “maximum time period”? Do you think that could cause confusion?

rustyrussell replied:
Agreed!  max_period_index is better, since we define "period index".

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+
+1. subtype: `recurrence_paywindow`
+2. data:
+    * [`u32`:`seconds_before`]
+    * [`tu32`:`seconds_after`]
```
Why is `seconds_before` in u32 (full-length) while `seconds_after` is in truncated u32?

shaavan replied:
Ah, I didn’t know about this, and it looks like this is common with every field. Will look into it more. Thanks!

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+   * which repeat from whenever you made the first payment
+
+Thus, each offer containing a recurring payment has:
+1. A `time_unit` defining 0 (seconds), 1 (days), 2 (months).
```
Why does `time_unit` have to be a single byte? Would we be able to represent a recurrence with an unusual composite interval, e.g., every 1 month + 15 days (Jan 1 → Feb 16 → Apr 1, etc.)? This doesn’t seem possible without allowing composite time units.

rustyrussell replied:
Agreed.  Also see: "Third Thursday of the Month" as things we don't support!

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+1. A `time_unit` defining 0 (seconds), 1 (days), 2 (months).
+2. A `period`, defining how often (in `time_unit`) it has to be paid.
+3. An optional `offer_recurrence_limit` of total payments to be paid.
+4. An optional `offer_recurrence_base`:
```
If `offer_recurrence_base` is optional, why is it assigned to field 26?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+   * The payer uses `invreq_recurrence_start` to indicate what period they are starting at.  If you don't want them to start at arbitrary periods, use `offer_absolute_expiry`.
+   * A `proportional_amount` flag: if set indicating that a payment made during the period itself will be charged proportionally to the remaining time in the period (e.g. 150 seconds into a 1500 second period gives a 10% discount).
+5. An optional `offer_recurrence_paywindow`:
+   * `seconds_before`, defining how many seconds prior to the start of the period a payment will be accepted.
+   * `seconds_after`, defining how many seconds after the start of the period a payment will be accepted.
```
Should we explicitly state whether `seconds_before` and `seconds_after` are inclusive? Is the range `[seconds_before, seconds_after]` or `(seconds_before, seconds_after)`?

Also, what if the period and range are large enough that expressing them in seconds is inefficient? For example, an offer with a 3-year period and a pay_window from 1 year before to 1 year after the payment date.

rustyrussell replied:
We do in fact define this in the actual requirements (Spoiler: it's inclusive!):

```markdown
      - if it includes `offer_recurrence_paywindow`:
        - MUST set `seconds_before` to the maximum number of seconds prior to a period for which it will accept payment or invoice_request for that period.
        - MUST set `seconds_after` to the maximum number of seconds into to a period for which it will accept payment or invoice_request for that period.
```

shaavan replied:
> I would guess that seconds_after would usually not be that high so it's OK to be "inefficient".

I see, that makes sense. Though I imagine it’s still a valid use case in some scenarios.

Also, defining a pay window as something like ±3 days feels more natural than specifying it as ±259,200 seconds. Do you think it might be more consistent (and readable) to express the pay window in terms of `(time_unit, window)`, similar to how `recurrence` is defined?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+   * `seconds_after`, defining how many seconds after the start of the period a payment will be accepted.
+  If this field is missing, payment will be accepted during the prior period and the paid-for period.
+
+Note that the `offer_absolute_expiry` field already covers the case where an offer is no longer valid after January 1st 2021.
```
I’m not sure if I understand the context this line is referencing to. It would be awesome if you can help me understand the reference here. Thanks a lot! :)

rustyrussell replied:
Updated:

Note: you can create an recurring offer with a fixed start which can only be used from the first period, using `offer_absolute_expiry`.  For example, my weekly book club starts January 1st 2026, for $5 a week: by setting `offer_absolute_expiry` to January 8th 2026, I ensure nobody can join after the first week.  Now, if only I could require them to actually read this week's book!

shaavan replied:
Looks good! (Also, wish there were a `book_reading_absolute_expiry` too 😄)

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+  - the start of period #0 is the `invoice_created_at`.`timestamp` of the first `invoice` for this particular offer and `invreq_payer_id`.
+
+To calculate UTC time of the start of period #`N` for `N` > 0:
+- if the period #0 start is on a leap second:
+  - the *period base* is one second before period #0 start.
```
I am curious as to why we chose “minus one second” over “plus one second”? Is it done so that, we maintain the month, date, hour, and minute of beginning of period, if the leap second was absent (For example, December 31st 23:59:60 to 23:59:59)?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -282,6 +373,40 @@ A writer of an offer:
+      - if it includes `offer_recurrence_paywindow`:
+        - MUST set `seconds_before` to the maximum number of seconds prior to a period for which it will accept payment or invoice_request for that period.
+        - MUST set `seconds_after` to the maximum number of seconds into to a period for which it will accept payment or invoice_request for that period.
+        - MAY NOT enforce this for the initial period for offers without `offer_recurrence_base`
```
Could this even be enforced? If `offer_recurrence_base` is not set, the period start-time isn’t defined until the first invoice is created. Wouldn’t that make the pay_window meaningless in this case? So should this be “MUST NOT” instead of “MAY NOT”?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -282,6 +373,40 @@ A writer of an offer:
+        - MUST set `seconds_before` to the maximum number of seconds prior to a period for which it will accept payment or invoice_request for that period.
+        - MUST set `seconds_after` to the maximum number of seconds into to a period for which it will accept payment or invoice_request for that period.
+        - MAY NOT enforce this for the initial period for offers without `offer_recurrence_base`
+        - SHOULD NOT set `seconds_after` to greater than the maximum number of seconds in a period.
```
What if it was set to be longer than the period, would this allow the user to pay for a period’s amount in the next period as well?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -439,14 +589,25 @@ while still allowing signature validation.
+    1. type: 93 (`invreq_recurrence_start`)
+    2. data:
+        * [`tu32`:`period_offset`]
+    1. type: 94 (`invreq_recurrence_cancel`)
```
Why is the data for type 92 (invreq_recurrence_cancel) not specified?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -457,6 +618,9 @@ The writer:
     - MUST set `signature`.`sig` as detailed in [Signature Calculation](#signature-calculation) using the `invreq_payer_id`.
+    - if it sets `invreq_recurrence_cancel`:
+      - SHOULD NOT send any future `invoice_request` for this offer with this `invreq_payer_id` without `invreq_recurrence_cancel` set.
+      - MAY omit `invreq_amount` and `invreq_quantity`.
```
Since `invreq_recurrence_cancel` is meant to indicate immediate cancellation (line 751: “MUST NOT send an invoice in reply”), shouldn’t `invreq_amount` and `invreq_quantity` be “MUST NOT present” as they would be meaningless in this context?

shaavan replied:
Okay, that makes sense! Keeping it flexible might reduce the transformation effort when adapting a regular invoice request into a recurrence cancellation.

shaavan commented on 12-offer-encoding.md:
```diff
@@ -471,6 +635,34 @@ The writer:
+        - MAY set `invreq_recurrence_cancel`.
+      - if `offer_recurrence_base` is present:
+        - MUST include `invreq_recurrence_start`
+        - MUST set `period_offset` to the period the sender wants for the initial request
```
Is the period offset from `offer_recurrence_base`? If so, could we clarify that? In general, would it make sense to have a dedicated section (like the recurrence section) defining the exact meaning of `invreq_recurrence_counter`, `invreq_recurrence_start`, and `invreq_recurrence_cancel`?

rustyrussell replied:
The offer_recurrence_base says "Period 0 always starts Feb 1st" for example.  Say it's monthly.

The user wants to start in August.  So, she sets `invreq_recurrence_start`'s `period_offset` to 6.  But her first invoice request, *as always* is for `invreq_recurrence_counter` 0.  That means it's trivial for the receiver to spot the first in a series of invoice requests....

I prefer to avoid nomenclature sections if we can make it clear in-situ.  Nobody (else!) ever reads them.

shaavan replied:
Okay, that clears it up! On a related note, do you think it would make sense to rename it to `period_index_offset`? Since it refers to the index rather than an actual time offset, it would better align with the `max_period_index` naming we landed on earlier, and could help avoid confusion.

shaavan commented on 12-offer-encoding.md:
```diff
@@ -740,6 +987,12 @@ A writer of an invoice:
       `version` as a valid witness version and `address` as a valid witness
       program
+  - if `offer_recurrence_optional` or `offer_recurrence_compulsory` are present:
+    - MUST set `invoice_recurrence_basetime`.`basetime` to the start of period #0 as calculated by [Period Calculation](#offer-period-calculation).
```
Here, it’s not specified whether this check applies to the first invoice, successive invoices, or all invoices. However, lines 1055–1056 clearly say to reject successive invoices where `invoice_recurrence_base` mismatches that of previous period’s invoice. This seems like a reader/writer instruction mismatch.

shaavan replied:
Got it, thanks for the clarification! So it’s the writer’s responsibility to always set `invoice_recurrence_base` to the correct basetime at period #0, and the reader’s job to ensure it stays consistent by comparing it with the previous invoice.

shaavan reviewed (commented):
Looks great on the first pass! 🚀
I’ve got a few questions and points I’d like to discuss. If some (or all 😄) have already been covered in earlier conversations I may have missed, please excuse the repetition. And in case I’ve misunderstood anything, please feel free to correct me. Thanks!

rustyrussell committed:
fixup!

Feedback from https://github.com/shaavan
1. rename max_period to max_period_index for a little more clarity.
2. clarify `offer_absolute_expiry` example.

shaavan reviewed (commented):
Thanks for all the clarifications, Rusty and Thomas!

shaavan commented on 12-offer-encoding.md:
```diff
@@ -282,6 +373,40 @@ A writer of an offer:
+    - MAY use `offer_recurrence_optional` if a payment of a single period would be useful (compatibility with pre-recurrence readers).
+    - MUST NOT use `offer_recurrence_optional` if `offer_recurrence_base` is present.
+  - if it includes either `offer_recurrence_optional` or `offer_recurrence_compulsory`:
+      - MUST set `time_unit` to 0 (seconds), 1 (days), 2 (months) or 3 (years).
```
`time_unit = 3 (years)` seems to have been removed from the spec—should we also remove it from the writer/reader instructions for consistency?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -282,6 +373,40 @@ A writer of an offer:
+    - MUST NOT use `offer_recurrence_optional` if `offer_recurrence_base` is present.
+  - if it includes either `offer_recurrence_optional` or `offer_recurrence_compulsory`:
+      - MUST set `time_unit` to 0 (seconds), 1 (days), 2 (months) or 3 (years).
+      - MUST set `period` to how often (in `time-unit`) it wants to be paid.
```
Should we add an explicit constraint for `period != 0` to avoid ambiguity? Something like:

Writer → MUST NOT set `period` to 0
Reader → MUST NOT respond to offers with `period` set to 0

rustyrussell committed:
Period feedback from https://github.com/shaavan

1. Remove reference to years (it's now seconds, days or months).
2. Make 0 illegal.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
fixup!

rustyrussell committed:
BOLT12: add stateless invoice support.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

shaavan commented on 12-offer-encoding.md:
```diff
@@ -480,6 +641,35 @@ The writer:
+        - MUST set `invreq_recurrence_counter` `counter` to 0.
+        - MUST NOT set `invreq_recurrence_cancel`.
+      - for any successive requests:
+        - MUST use the same `invreq_payer_id` as the initial request.
```
Why does the payer need to maintain a stable `invreq_payer_id` across recurrence invoice requests? This appears to introduce public linkability for the payer across the lifetime of a recurrence session.

If the goal is to allow the payee to associate invoice requests with a particular recurrence session, could the recently added `invoice_recurrence_prev_state` and `invoice_recurrence_next_state` fields serve that purpose instead, avoiding the need for a stable payer identifier?

shaavan commented on 12-offer-encoding.md:
Recurrence sessions may be very long-lived, potentially lasting months or years.

When `offer_paths` is present, subsequent recurrence invoice requests are required to use the same blinded paths from the original offer. This appears to require those paths to remain usable for the entire lifetime of the recurrence session, potentially long after the original offer itself has expired.

If those paths become invalid over time, how is the payer expected to continue the recurrence session?

More generally, should long-lived recurrence sessions depend on communication paths established at offer creation time, or is some mechanism for path refresh expected?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+1. A `time_unit` defining 0 (seconds), 1 (days), 2 (months).
+2. A `period`, defining how often (in `time_unit`) it has to be paid.
+3. An optional `offer_recurrence_limit` of total payments to be paid.
+4. An optional `offer_recurrence_base`:
```
The current design allows a recurrence to be anchored either via `offer_recurrence_base` in the offer or via `invoice_recurrence_basetime` in the first invoice.

What advantages does allowing recurrence anchoring to originate from either the offer or the first invoice provide compared to requiring `period #0` to always be defined at offer creation time and making the offer the sole source of truth for recurrence anchoring?

### Additional Thoughts

I can see the value of payer-driven recurrence basetimes. However, I am curious whether that flexibility could be moved to the application layer while keeping recurrence anchoring entirely offer-defined.

For example:

* A merchant could support generating recurrence-on-demand offers, with the basetime chosen when the payer decides to begin the recurrence.
* For a long-lived static offer, a payer could begin participating after several recurrence periods have already elapsed, with the recurrence offset indicating the period at which participation began.

Would an approach like this preserve the same flexibility while keeping a single source of truth for recurrence anchoring?

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+
+Note: you can create an recurring offer with a fixed start which can only be used from the first period, using `offer_absolute_expiry`.  For example, my weekly book club starts January 1st 2026, for $5 a week: by setting `offer_absolute_expiry` to January 8th 2026, I ensure nobody can join after the first week.  Now, if only I could require them to actually read this week's book!
+
+For wallets which don't (yet) support recurrence, the basic recurrence field comes in two variants: `offer_recurrence_compulsory` if they should not attempt a payment, and `offer_recurrence_optional` if it still makes sense for them to attempt a single payment.
```
The current design distinguishes between recurrence being optional or compulsory at the protocol level.

What advantages are gained by encoding this distinction in the recurrence specification itself?

Could recurrence instead be treated as a protocol extension where recurrence is always optional at the spec level, with applications advertising and deciding whether one-time payments are accepted or rejected?

jharveyb commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+   * which repeat from whenever you made the first payment
+
+Thus, each offer containing a recurring payment has:
+1. A `time_unit` defining 0 (seconds), 1 (days), 2 (months).
```
From previous discussion: the smallest 'reasonable' unit here is likely hours, not seconds (for example, if mobile wallets need to wake up).

benthecarman replied:
yes I think we explicitly want this to be larger, otherwise people might attempt to use this for streaming payments which seems incorrect / a misuse

benthecarman commented on 12-offer-encoding.md:
```diff
@@ -532,11 +723,45 @@ The reader:
         - if `offer_currency` is not the `invreq_chain` currency, convert to the
           `invreq_chain` currency.
         - if `invreq_quantity` is present, multiply by `invreq_quantity`.`quantity`.
+        - if `offer_recurrence_base` is present and `proportional_amount` is 1:
```
it's not defined for what to do if `offer_recurrence_base` is not present when `proportional_amount` is. I assume this is invalid?

jharveyb commented on 12-offer-encoding.md:
```diff
@@ -283,6 +373,41 @@ A writer of an offer:
+        - MUST set `seconds_before` to the maximum number of seconds prior to a period for which it will accept payment or invoice_request for that period.
+        - MUST set `seconds_after` to the maximum number of seconds into to a period for which it will accept payment or invoice_request for that period.
+        - MAY NOT enforce this for the initial period for offers without `offer_recurrence_base`
+        - SHOULD NOT set `seconds_after` to greater than the maximum number of seconds in a period.
```
Related to previous comments here: should `seconds_before` also be clamped to the size of a period? Such that a payer can only pay in advance for a maximum of one recurrence.

Without clamping, a payer could pre-pay for a full recurrence (or a significant portion of it) at once, which may be undesirable behavior from the perspective of the payee. From what I've seen of subscriptions in the wild, usually the price of a recurrence is tied to the period, and making multiple pre-payments is disallowed.

Concrete example:

A monthly subscription of 10 USD, for 1 year. `seconds_before` = 1 year, `seconds_after` = 1 month.

IIUC, with the spec as written the payer can pay for all 12 periods upfront, effectively 'locking in' the current exchange rate. I would *expect* the payee here to want to provide a different offer for a 12 month prepayment, likely with a lower price (100 USD for 1 year of service, or 100 USD per year, etc.)

tl;dr not clamping BOTH `seconds_before` and `seconds_after` seems like leaving a potential footgun around. Happy to hear counterarguments though.

shaavan commented on 12-offer-encoding.md:
```diff
@@ -229,6 +229,97 @@ The human-readable prefix for offers is `lno`.
+    * [`byte`:`time_unit`]
+    * [`tu32`:`period`]
+
+1. subtype: `recurrence_base`
+2. data:
+    * [`byte`:`proportional_amount`]
+    * [`tu64`:`basetime`]
```
I was wondering whether `basetime` and `proportional_amount` should be separate subtypes?

For example, a merchant may want to let the payer choose when to start a recurrence (i.e., by not setting `basetime`) while also allowing them to pay a proportional amount beginning with the first successive period.

Also, I think @jharveyb raised an interesting point (offline) about proportional amounts. IIUC, the current interpretation effectively acts as a discount. For example, if you pay 7 days late in a month-long period, you only pay for the remaining ~75% of the period.

In practice, though, merchants often don't terminate service immediately when a payment is late. For example, Spotify gives users a grace period before suspending service.

Jonathan's suggestion, if I understood it correctly, was that proportional amounts could instead be interpreted as a late-payment penalty. In other words, the payer would owe *more* if they paid after the due date, rather than less.

That might also help address another issue. Today, once the payment window is missed, the recurrence becomes void. If proportional amounts represented a penalty instead, it might allow payments even after the payment window, with the additional amount serving as the late-payment penalty.

I'd love to hear @jharveyb's thoughts on whether that's what he had in mind. Thanks!
